src/charm.py

```diff
@@ -39,10 +41,124 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.on.get_primary_action, self._on_get_primary)
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.get_initial_password_action, self._on_get_initial_password)
         self._cluster = Patroni(self._unit_ip)
 
+    def _on_get_primary(self, event: ActionEvent) -> None:
+        """Get primary instance."""
+        try:
+            primary = self._patroni.get_primary(unit_name_pattern=True)
+            event.set_results({"primary": primary})
+        except RetryError as e:
+            logger.error(f"failed to get primary with error {e}")
+
+    @retry(
+        stop=stop_after_delay(60),
+        wait=wait_fixed(5),
+        reraise=True,
+    )
+    def _change_primary(self) -> None:
+        """Change the primary member of the cluster."""
+        # Inform the first of the remaining available members to not incur the risk
+        # of triggering a switchover to a member that is also being removed.
+        if not self._patroni.are_all_members_ready():
+            logger.info("could not switchover because not all members are ready")
+            return
+
+        # Try switchover and raise and exception if it doesn't succeed.
+        # If it doesn't happen on time, Patroni will automatically run a fail-over.
+        try:
+            self._patroni.switchover()
+            logger.info("successful switchover")
+        except SwitchoverFailedError as e:
+            logger.error(f"switchover failed with reason: {e}")
+
+    def _get_ip_by_unit(self, unit: Unit) -> str:
+        """Get the IP address of a specific unit."""
+        return self._peers.data[unit].get("private-address")
+
+    @property
+    def _hosts(self) -> set:
+        """List of the current Juju hosts.
+
+        Returns:
+            a set containing the current Juju hosts
+                with the names in the k8s pod name format
```

Comment:
remove k8s

Reply:
Good catch. Fixed in https://github.com/canonical/postgresql-operator/pull/10/commits/61273256b0210debc93689b8cf62157923190cd6.